src/drivers/legos/touchpad_driver.rs

```diff
@@ -0,0 +1,170 @@
+    pub fn poll(&mut self) -> Result<Vec<Event>, Box<dyn Error + Send + Sync>> {
+        // Read data from the device into a buffer
+        let mut events = vec![];
+        let mut buf = [0; TOUCH_PACKET_SIZE];
+        let bytes_read = self.device.read_timeout(&mut buf[..], HID_TIMEOUT)?;
+        let report_id = buf[0];
+
+        if bytes_read == TOUCH_PACKET_SIZE && report_id == TOUCH_REPORT_ID {
+            events = match self.handle_touch_report(buf) {
+                Ok(events) => events,
+                Err(e) => {
+                    log::error!("Got error processing TouchInputDataReport: {e:?}");
+                    vec![]
+                }
+            };
+        }
+
+        // There is no release event, so check to see if we are still touching.
+        if self.is_touching && (self.last_touch.elapsed() > Duration::from_millis(25)) {
+            self.is_touching = false;
+            events.push(Event::Axis(AxisEvent::Touchpad(TouchAxisInput {
+                index: 0,
+                is_touching: false,
+                x: 0,
+                y: 0,
+            })));
+        };
+
+        Ok(events)
+    }
+
+    /* Touchpad */
+    /// Unpacks the buffer into a [TouchInputDataReport] structure and updates
+    /// the internal touch_state.
+    fn handle_touch_report(
+        &mut self,
+        buf: [u8; TOUCH_PACKET_SIZE],
+    ) -> Result<Vec<Event>, Box<dyn Error + Send + Sync>> {
+        let input_report = TouchpadDataReport::unpack(&buf)?;
+
+        // Print input report for debugging
+        //log::debug!("--- Input report ---");
+        //log::debug!("{input_report}");
+        //log::debug!(" ---- End Report ----");
+
+        // Update the state
+        let old_state = self.update_touchpad_state(input_report);
+
+        // Translate the state into a stream of input events
+        let events = self.translate_touch(old_state);
+
+        Ok(events)
+    }
+
+    /// Update touchinput state
+    fn update_touchpad_state(
+        &mut self,
+        input_report: TouchpadDataReport,
+    ) -> Option<TouchpadDataReport> {
+        let old_state = self.touchpad_state;
+        self.touchpad_state = Some(input_report);
+        old_state
+    }
+
+    /// Translate the state into individual events
+    fn translate_touch(&mut self, old_state: Option<TouchpadDataReport>) -> Vec<Event> {
+        let mut events = Vec::new();
+        let Some(state) = self.touchpad_state else {
+            return events;
+        };
+
+        // Translate state changes into events if they have changed
+        let Some(old_state) = old_state else {
+            return events;
+        };
+
+        // Binary events
+        if state.pressed != old_state.pressed {
+            events.push(Event::Button(ButtonEvent::RPadPress(BinaryInput {
+                pressed: state.pressed,
+            })));
+            events.push(Event::Trigger(TriggerEvent::RpadForce(TriggerInput {
+                value: 32 * state.pressed as u8,
```

Comment:
Is there a reason to multiply this by 32?

Reply:
Pressed is a bool. I'm emulating about 1/4 max pressure which, from my testing, is about what a normal couch on the deck is. There's no pressure sensor on the go s.